rules/onchain_budget_test.go

```diff
@@ -0,0 +1,281 @@
+	// amount should be updated.
+	_, err = enf.HandleRequest(
+		ctx, "/lnrpc.Lightning/OpenChannel", &lnrpc.OpenChannelRequest{
+			SatPerVbyte:        10,
+			LocalFundingAmount: 20,
+			PushSat:            10,
+		},
+	)
+	require.NoError(t, err)
+	assertSpentAmt(0, 30)
+
+	// If the above request causes an error on the lnd side for whatever
+	// reason, the pending amounts should be reset.
+	_, err = enf.HandleErrorResponse(
+		ctx, "/lnrpc.Lightning/OpenChannel", errors.New("no good"),
+	)
+	require.NoError(t, err)
+	assertSpentAmt(0, 0)
+
+	// Try the call again.
+	_, err = enf.HandleRequest(
+		ctx, "/lnrpc.Lightning/OpenChannel", &lnrpc.OpenChannelRequest{
+			SatPerVbyte:        10,
+			LocalFundingAmount: 20,
+			PushSat:            10,
+		},
+	)
+	require.NoError(t, err)
+	assertSpentAmt(0, 30)
+
+	// A channel pending response will not move the pending amount from
+	// above into the spent amount.
+	_, err = enf.HandleResponse(
+		ctx, "/lnrpc.Lightning/OpenChannel", &lnrpc.OpenStatusUpdate{
+			Update: &lnrpc.OpenStatusUpdate_ChanPending{},
+		},
+	)
+	require.NoError(t, err)
+	assertSpentAmt(0, 30)
+
+	// Now check that a response using the same request ID will move the
+	// pending amount from above into the spent amount.
+	_, err = enf.HandleResponse(
+		ctx, "/lnrpc.Lightning/OpenChannel", &lnrpc.OpenStatusUpdate{
+			Update: &lnrpc.OpenStatusUpdate_ChanOpen{},
+		},
+	)
+	require.NoError(t, err)
+	assertSpentAmt(30, 0)
+
+	// There is now 70 sats left to spend.
+	// Assert that trying to do a batch open where each individual channel
+	// amount is ok but the total exceeds the budget will fail.
+	_, err = enf.HandleRequest(
+		ctx, "/lnrpc.Lightning/BatchOpenChannel",
+		&lnrpc.BatchOpenChannelRequest{
+			SatPerVbyte: 10,
+			Channels: []*lnrpc.BatchOpenChannel{
+				{
+					LocalFundingAmount: 20,
+					PushSat:            10,
+				},
+				{
+					LocalFundingAmount: 20,
+					PushSat:            10,
+				},
+				{
+					LocalFundingAmount: 20,
+					PushSat:            10,
+				},
+			},
+		},
+	)
+	require.Error(t, err)
+	assertSpentAmt(30, 0)
+
+	// Try again with a valid batch.
+	_, err = enf.HandleRequest(
+		ctx, "/lnrpc.Lightning/BatchOpenChannel",
+		&lnrpc.BatchOpenChannelRequest{
+			SatPerVbyte: 10,
+			Channels: []*lnrpc.BatchOpenChannel{
+				{
+					LocalFundingAmount: 20,
+					PushSat:            10,
+				},
+				{
+					LocalFundingAmount: 20,
+					PushSat:            10,
+				},
+			},
+		},
+	)
+	require.NoError(t, err)
+	assertSpentAmt(30, 60)
+
+	// Before confirming the above request with a response, lets create a
+	// new request but with a different request id. First ensure that an
+	// amount that would exceed the current budget would fail.
+	cfg.ReqID = 2
+	_, err = enf.HandleRequest(
+		ctx, "/lnrpc.Lightning/OpenChannel", &lnrpc.OpenChannelRequest{
+			LocalFundingAmount: 15,
+		},
+	)
+	require.Error(t, err)
+	assertSpentAmt(30, 60)
```

Comment:
nit:
I think it maybe would make sense to prove here that if the previous requests that keeps `60` of the `pending_balance` fails, this or a similar request succeeds when attempted once more.
That's just to prove that it **_is_** intended behaviour that when processing concurrent requests the order of the requests does actually matter for the outcome.

Feel free to ignore this if you think this is unnecessary coverage though :)!

Reply:
Thanks for thinking about these edge cases :pray:! I have added a test in the end (also noted that there could be issues with accounting if the URI was changed for errors), hope I understood you correctly concerning the edge case.